hudi-common/src/main/java/org/apache/hudi/common/model/HoodieAvroRecordMerger.java

```diff
@@ -71,10 +71,10 @@ public HoodieRecordType getRecordType() {
     return HoodieRecordType.AVRO;
   }
 
-  private HoodieRecord preCombine(HoodieRecord older, HoodieRecord newer) {
-    HoodieRecordPayload picked = unsafeCast(((HoodieAvroRecord) newer).getData().preCombine(((HoodieAvroRecord) older).getData()));
-    if (picked instanceof HoodieMetadataPayload) {
-      // NOTE: HoodieMetadataPayload return a new payload
+  private HoodieRecord preCombine(HoodieRecord older, HoodieRecord newer, Schema schema, Properties props) {
+    HoodieRecordPayload picked = unsafeCast(((HoodieAvroRecord) newer).getData().preCombine(((HoodieAvroRecord) older).getData(), schema, props));
+    if (picked instanceof HoodieMetadataPayload || picked instanceof OverwriteNonDefaultsWithLatestAvroPayload) {
+      // NOTE: When payload class returns a new payload instead of choosing from older/newer payload, create a new record.
       return new HoodieAvroRecord(newer.getKey(), picked, newer.getOperation());
     }
     return picked.equals(((HoodieAvroRecord) newer).getData()) ? newer : older;
```

Comment:
I think we can nuke this optimization avoiding creation of the new record and just always create it in that case (to make sure this doesn't affect all random payloads out there)